src/api/procedures/linkCaDocs.ts

```diff
@@ -0,0 +1,87 @@
+import BigNumber from 'bignumber.js';
+import { differenceWith } from 'lodash';
+import { DocumentId, TxTags } from 'polymesh-types/types';
+
+import { PolymeshError, Procedure, SecurityToken } from '~/internal';
+import { ErrorCode, RoleType, TokenDocument } from '~/types';
+import { ProcedureAuthorization } from '~/types/internal';
+import { documentToTokenDocument, stringToTicker } from '~/utils/conversion';
+import { documentComparator } from '~/utils/internal';
+
+export interface LinkCaDocsParams {
+  documents: TokenDocument[];
+}
+
+/**
+ * @hidden
+ */
+export type Params = LinkCaDocsParams & {
+  id: BigNumber;
+  ticker: string;
+};
+
+/**
+ * @hidden
+ */
+export async function prepareLinkCaDocs(
+  this: Procedure<Params, void>,
+  args: Params
+): Promise<void> {
+  const {
+    context: {
+      polymeshApi: {
+        tx: { corporateAction },
+        query: {
+          asset: { assetDocuments },
+        },
+      },
+    },
+    context,
+  } = this;
+  const { id: caId, ticker, documents } = args;
+
+  const rawAssetDocuments = await assetDocuments.entries(stringToTicker(ticker, context));
+
+  const currentAssetDocIds: DocumentId[] = [];
+  const currentAssetDocs: TokenDocument[] = [];
+
+  rawAssetDocuments.forEach(([key, doc]) => {
+    const [, id] = key.args;
+    currentAssetDocIds.push(id);
+    currentAssetDocs.push(documentToTokenDocument(doc));
+  });
```

Comment:
This approach results in every document associated to the asset being linked to the CA because you're not filtering anything. What you need to do is:

- rename `currentAssetDocIds` to `docIdsToLink`
- use lodash `remove` (together with `isEqual`) to search if the doc (converted to `TokenDocument`) is present in the input array (this will also remove the element from the input array so that subsequent searches are more performant). Be mindful that the result of `remove` is an array, so check for length
- if found, push the id to the `docIdsToLink` array, otherwise do nothing
- if there are any elements left in `documents` after the loop, it means the array had documents not in the asset, so you can throw the error (also, add `documents` to the error data and test for it)